File: libs/power-log-parser/src/lib/state/parser-state.ts
import { CardType, GameTag, GameType, ScenarioId, Zone } from '@firestone-hs/reference-data';
import {
	BaseEntity,
	ChangeEntity,
	ChosenEntities,
	Choices,
	FullEntity,
	Game,
	GameEntity,
	HearthstoneReplay,
	IEntityData,
	MetaData,
	Node,
	NodeType,
	Option,
	Options,
	Player,
	PlayerEntity,
	SendChoices,
	ShowEntity,
	SubSpell,
	Tag,
} from '../models';
import { GameEventProvider } from '../game-event';
import { Logger } from '../logger';
import { GameState } from './game-state';
import type { StateFacade } from './state-facade';

export enum StateType {
	GameState = 'GameState',
	PowerTaskList = 'PowerTaskList',
}

export interface INodeParser {
	NewNode(node: Node, stateType: StateType): void;
	CloseNode(node: Node, stateType: StateType): void;
	EnqueueGameEvent(providers: GameEventProvider[]): void;
	ClearQueue(): void;
	Reset(state: ParserState, helper: StateFacade): void;
}

export class ParserState {
	GameState: GameState = new GameState();

	NodeParser!: INodeParser;
	StateFacade!: StateFacade;
	Replay!: HearthstoneReplay;
	CurrentGame!: Game;
	GameData: any | null = null;
	SendChoices: SendChoices | null = null;
	Choices: Choices | null = null;
	Options: Options | null = null;
	CurrentOption: Option | null = null;
	LastOption: any | null = null;
	CurrentSubSpell: SubSpell | null = null;
	FirstPlayerEntityId: number = -1;
	CurrentPlayerId: number = -1;
	CurrentChosenEntites: ChosenEntities | null = null;
	Ended: boolean = false;
	NumberOfCreates: number = 0;
	ReconnectionOngoing: boolean = false;
	Spectating: boolean = false;

	StateType: StateType;

	private _node: Node | null = null;
	get Node(): Node | null {
		return this._node;
	}
	set Node(value: Node | null) {
		if (value !== this._node) {
			if (
				this._node != null &&
				(this._node.Type === NodeType.FullEntity ||
					this._node.Type === NodeType.ShowEntity ||
					this._node.Type === NodeType.ChangeEntity)
			) {
				this.EndAction();
				if (this._node.Type === NodeType.ShowEntity) {
					this.GameState.ShowEntity(this._node.Object as ShowEntity);
				} else if (this._node.Type === NodeType.FullEntity) {
					this.GameState.FullEntity(this._node.Object as FullEntity, false);
				} else if (this._node.Type === NodeType.ChangeEntity) {
					this.GameState.ChangeEntity(this._node.Object as ChangeEntity);
				}
			}
			if (this._node != null && this._node.Type === NodeType.PlayerEntity) {
				this.NodeParser.CloseNode(this._node, this.StateType);
				this.GameState.PlayerEntity(this._node.Object as PlayerEntity);
			}
			if (this._node != null && this._node.Type === NodeType.GameEntity) {
				this.NodeParser.CloseNode(this._node, this.StateType);
				this.GameState.GameEntity(this._node.Object as GameEntity);
			}
			if (this._node != null && this._node.Type === NodeType.MetaData) {
				this.EndAction();
			}
			this._node = value;
		}
	}

	private _localPlayer: Player | null = null;
	get LocalPlayer(): Player | null {
		return this._localPlayer;
	}

	private _opponentPlayer: Player | null = null;
	get OpponentPlayer(): Player | null {
		return this._opponentPlayer;
	}

	constructor(type: StateType, nodeParser: INodeParser, stateFacade: StateFacade) {
		Logger.Log('Calling reset from ParserState constructor', type);
		this.StateType = type;
		this.StateFacade = stateFacade;
		this.NodeParser = nodeParser;
	}

	SetLocalPlayer(value: Player, timestamp: string, data: string, sendEvent: boolean): void {
		this._localPlayer = value;
		value.IsMainPlayer = true;
		const playerEntity = this.getPlayers().find((player) => player.PlayerId === value.PlayerId);
		if (playerEntity) {
			playerEntity.IsMainPlayer = value.IsMainPlayer;
		}
		if (sendEvent) {
			const localPlayer = this._localPlayer;
			this.NodeParser.EnqueueGameEvent([
				GameEventProvider.Create(
					timestamp,
					'LOCAL_PLAYER',
					() => ({
						Type: 'LOCAL_PLAYER',
						Value: localPlayer,
					}),
				false,
				new Node(NodeType.Placeholder, null, 0, null, data),
			),
		]);
	}
}

	SetOpponentPlayer(value: Player, timestamp: string, data: string, sendEvent: boolean): void {
		this._opponentPlayer = value;
		value.IsMainPlayer = false;
		const playerEntity = this.getPlayers().find((player) => player.PlayerId === value.PlayerId);
		if (playerEntity) {
			playerEntity.IsMainPlayer = value.IsMainPlayer;
		}
		if (sendEvent) {
			const opponentPlayer = this._opponentPlayer;
			this.NodeParser.EnqueueGameEvent([
				GameEventProvider.Create(
					timestamp,
					'OPPONENT_PLAYER',
					() => ({
						Type: 'OPPONENT_PLAYER',
						Value: {
							OpponentPlayer: opponentPlayer,
						},
					}),
					false,
					new Node(NodeType.Placeholder, null, 0, null, data),
				),
			]);
		}
	}

	Reset(helper: StateFacade): void {
		this.GameState.Reset(this);
		this.NodeParser.Reset(this, helper);
		this.Replay = new HearthstoneReplay();
		this.Replay.Games = [];
		this.CurrentGame = new Game();
		this._localPlayer = null;
		this._opponentPlayer = null;
		this._node = null;
		this.Node = null;
		this.GameData = null;
		this.SendChoices = null;
		this.Choices = null;
		this.Options = null;
		this.CurrentOption = null;
		this.LastOption = null;
		this.FirstPlayerEntityId = -1;
		this.CurrentPlayerId = -1;
		this.CurrentChosenEntites = null;
		this.Ended = false;
		this.ReconnectionOngoing = false;
		this.NumberOfCreates = 0;
		Logger.Log(`resetting game state, ended is now Ended=${this.Ended}`, this.StateType);

		this._isBattlegrounds = null;
		this._cachedPlayers = [];
	}

	PartialReset(): void {
		const gameEntity = this.CurrentGame.Data.find((d) => d instanceof GameEntity) as GameEntity | undefined;
		const playerEntities = this.getPlayers();
		const keysToKeep = new Set<number>([gameEntity!.Id]);
		for (const player of playerEntities) {
			keysToKeep.add(player.Id);
		}
		const filteredEntities = new Map<number, FullEntity>();
		for (const [key, value] of this.GameState.CurrentEntities) {
			if (keysToKeep.has(key)) {
				filteredEntities.set(key, value);
			}
		}
		this.GameState.CurrentEntities = filteredEntities;
	}

	CreateNewNode(newNode: Node): void {
		this.NodeParser.NewNode(newNode, this.StateType);
	}

	EndAction(): void {
		if (this.Node!.Type !== NodeType.Game) {
			this.NodeParser.CloseNode(this.Node!, this.StateType);
		}
	}

	EndCurrentGame(): void {
		this.Ended = true;
		Logger.Log(`Game Ended in type=${this.StateType}`, `Ended=${this.Ended}`);
	}

	UpdateCurrentNode(...types: NodeType[]): void {
		while (this.Node?.Parent != null && types.every((x) => x !== this.Node!.Type)) {
			this.Node = this.Node.Parent;
		}
	}

	private _cachedPlayers: PlayerEntity[] = [];
	getPlayers(): PlayerEntity[] {
		if (this._cachedPlayers.length > 0) {
			return this._cachedPlayers;
		}
		const players: PlayerEntity[] = [];
		const dataCopy = [...this.CurrentGame.Data];
		for (const x of dataCopy) {
			if (x instanceof PlayerEntity) {
				players.push(x);
			}
		}
		return players;
	}

	GetPlayerForController(controllerId: number): PlayerEntity | null {
		const players = this.getPlayers();
		for (const player of players) {
			if (player.PlayerId === controllerId) return player;
		}
		return null;
	}

	/** True once DebugPrintGame has assigned a non-blank name (bots with AccountHi 0 skip the wait). */
	private playerHasDebugName(player: PlayerEntity): boolean {
		if (player.AccountHi === '0') {
			return true;
		}
		return player.Name != null && player.Name.trim().length > 0;
	}

	TryAssignLocalPlayer(timestamp: string, data: string): void {
		if (this.IsMercenaries()) {
			if (this.getPlayers().length === 3 && this.CurrentGame.ScenarioID === (ScenarioId.LETTUCE_PVP as number)) {
				this.CurrentGame.ScenarioID = ScenarioId.LETTUCE_PVP_VS_AI as number;
			}

			let localPlayerPlayerId = -1;
			let opponentPlayerPlayerId = -1;

			if (this.IsMercenariesPvE() && this.getPlayers().length === 3) {
				localPlayerPlayerId = this.getPlayers()[2].PlayerId;
				opponentPlayerPlayerId = this.getPlayers()[1].PlayerId;
			} else {
				const allFullEntities = this.CurrentGame
					.FilterGameData(FullEntity)
					.filter((d): d is FullEntity => d instanceof FullEntity);

				const localMerc = allFullEntities.find(
					(d) => d.GetTag(GameTag.LETTUCE_MERCENARY) === 1 && (d.CardId?.length ?? 0) > 0,
				);
				localPlayerPlayerId = localMerc?.GetEffectiveController() ?? -1;

				const opponentMerc = allFullEntities.find(
					(d) =>
						d.GetTag(GameTag.LETTUCE_MERCENARY) === 1 &&
						((d.CardId?.length ?? 0) === 0 || d.GetZone() === (Zone.PLAY as number)) &&
						d.GetEffectiveController() !== localPlayerPlayerId,
				);
				opponentPlayerPlayerId = opponentMerc?.GetEffectiveController() ?? -1;
			}

			for (const player of this.getPlayers()) {
				if (player.PlayerId === opponentPlayerPlayerId && data.includes('PlayerID=' + opponentPlayerPlayerId)) {
					const newPlayer = Player.from(player);
					this.SetOpponentPlayer(newPlayer, timestamp, data, this.StateType === StateType.GameState);
					return;
				} else if (
					player.PlayerId === localPlayerPlayerId &&
					data.includes('PlayerID=' + localPlayerPlayerId)
				) {
					const newPlayer = Player.from(player);
					this.FirstPlayerEntityId = player.Id;
					this.SetLocalPlayer(newPlayer, timestamp, data, this.StateType === StateType.GameState);
					return;
				}
			}
			return;
		}

		if (this.LocalPlayer != null && this.OpponentPlayer != null) {
			return;
		}

		for (const player of this.getPlayers()) {
			if (!this.playerHasDebugName(player)) {
				return;
			}
			const playerEntityIdTag = player.Tags.find((t) => t.Name === (GameTag.HERO_ENTITY as number));
			if (playerEntityIdTag == null) {
				return;
			}
		}

		let showEntities: IEntityData[] = this.CurrentGame
			.FilterGameData(ShowEntity)
			.filter((d): d is ShowEntity => d instanceof ShowEntity)
			.filter((d) => d.GetZone() === (Zone.HAND as number));

		if (showEntities.length === 0) {
			Logger.Log('No show entity, fallback to fullentity in hand', '');
			showEntities = this.CurrentGame
				.FilterGameData(FullEntity)
				.filter((d): d is FullEntity => d instanceof FullEntity)
				.filter((d) => d.GetZone() === (Zone.HAND as number))
				.filter((d) => d.GetTag(GameTag.CREATOR) !== this.GameState.GetGameEntity()?.Id);

			if (showEntities.length === 0) {
				Logger.Log('No full entity in hand, fallback to fullentity', '');
				showEntities = this.CurrentGame.FilterGameData(FullEntity) as unknown as IEntityData[];
			}
		}

		for (const entity of showEntities) {
			if (
				entity.CardId != null &&
				entity.CardId.length > 0 &&
				this.GetTagFromList(entity.Tags, GameTag.CARDTYPE) !== (CardType.ENCHANTMENT as number) &&
				this.GetTagFromList(entity.Tags, GameTag.ZONE) !== (Zone.DECK as number)
			) {
				const entityId = entity.Entity;
				const fullEntity = this.GetEntity(entityId);
				if (!fullEntity) continue;
				const controllerId = fullEntity.GetEffectiveController();
				for (const player of this.getPlayers()) {
					if (player.GetEffectiveController() === controllerId) {
						const newPlayer = Player.from(player);
						const playerEntityIdTag = player.Tags.find(
							(t) => t.Name === (GameTag.HERO_ENTITY as number),
						)!;
						const playerEntity = this.CurrentGame.Data.filter(
							(d): d is FullEntity => d instanceof FullEntity,
						).find((e) => e.Id === playerEntityIdTag.Value)!;
						newPlayer.CardID = playerEntity.CardId;
						this.SetLocalPlayer(newPlayer, timestamp, data, this.StateType === StateType.GameState);
					}
				}
				if (this.LocalPlayer != null) {
					for (const player of this.getPlayers()) {
						if (player.Id === this.LocalPlayer.Id) {
							continue;
						}
						const newPlayer = Player.from(player);
						const playerEntityIdTag = player.Tags.find(
							(t) => t.Name === (GameTag.HERO_ENTITY as number),
						)!;
						const playerEntity = this.CurrentGame.Data.filter(
							(d): d is FullEntity => d instanceof FullEntity,
						).find((e) => e.Id === playerEntityIdTag.Value)!;
						newPlayer.CardID = playerEntity.CardId;
						this.SetOpponentPlayer(newPlayer, timestamp, data, this.StateType === StateType.GameState);
					}
					return;
				}
			}
		}

		if (this._localPlayer == null && this._opponentPlayer == null) {
			Logger.Log(
				'ERROR TO LOG: Could not assign local player ' + data,
				this.getPlayers()
					?.map((player) => player.Name)
					.join(', ') ?? '',
			);
		}
	}

	GetTagFromList(tags: Tag[], tag: GameTag): number {
		const ret = tags.find((t) => t.Name === (tag as number));
		return ret == null ? -1 : ret.Value;
	}

	GetEntity(id: number): BaseEntity | undefined {
		return this.CurrentGame
			.FilterGameData(FullEntity, PlayerEntity)
			.filter((data): data is BaseEntity => data instanceof BaseEntity)
			.find((e) => e.Id === id);
	}

	private _isBattlegrounds: boolean | null = null;
	IsBattlegrounds(): boolean {
		if (this._isBattlegrounds != null) {
			return this._isBattlegrounds;
		}
		if (this.CurrentGame.GameType === -1) {
			return false;
		}
		this._isBattlegrounds =
			this.CurrentGame.GameType === (GameType.GT_BATTLEGROUNDS as number) ||
			this.CurrentGame.GameType === (GameType.GT_BATTLEGROUNDS_FRIENDLY as number) ||
			this.CurrentGame.GameType === (GameType.GT_BATTLEGROUNDS_DUO as number) ||
			this.CurrentGame.GameType === (GameType.GT_BATTLEGROUNDS_DUO_FRIENDLY as number);
		return this._isBattlegrounds;
	}

	IsBattlegroundsDuos(): boolean {
		return (
			this.CurrentGame.GameType === (GameType.GT_BATTLEGROUNDS_DUO as number) ||
			this.CurrentGame.GameType === (GameType.GT_BATTLEGROUNDS_DUO_FRIENDLY as number)
		);
	}

	InRecruitPhase(): boolean {
		return this.GameState.GetGameEntity()!.GetTag(GameTag.BOARD_VISUAL_STATE) === 1;
	}

	InCombatPhase(): boolean {
		return this.GameState.GetGameEntity()!.GetTag(GameTag.BOARD_VISUAL_STATE) === 2;
	}

	IsMercenaries(): boolean {
		return (
			this.IsMercenariesPvE() ||
			[GameType.GT_MERCENARIES_PVP as number].includes(this.CurrentGame.GameType)
		);
	}

	IsMercenariesPvE(): boolean {
		return [
			GameType.GT_MERCENARIES_AI_VS_AI as number,
			GameType.GT_MERCENARIES_FRIENDLY as number,
			GameType.GT_MERCENARIES_PVE as number,
			GameType.GT_MERCENARIES_PVE_COOP as number,
		].includes(this.CurrentGame.GameType);
	}

	IsReconnecting(gameSeedForCurrentLogs: number): boolean {
		Logger.Log(
			'Is reconnecting?',
			`statType=${this.StateType}, ended=${this.Ended}, creates=${this.NumberOfCreates}, spectating=${this.Spectating}, gameSeed=${this.CurrentGame?.GameSeed}, currentGameSeed=${gameSeedForCurrentLogs}`,
		);
		return (
			gameSeedForCurrentLogs !== 0 &&
			this.CurrentGame != null &&
			this.CurrentGame.GameSeed !== 0 &&
			this.CurrentGame.GameSeed === gameSeedForCurrentLogs
		);
	}

	ClearActiveSubSpell(): void {
		let current = this.CurrentSubSpell;
		let parent: SubSpell | null = null;
		if (current?.Spell == null) {
			this.CurrentSubSpell = null;
		}
		while (current?.Spell != null) {
			parent = current;
			current = current.Spell;
		}
		if (parent != null) {
			parent.Spell = null;
		}
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Parser/ParserState.cs
#region

using System;
using System.Threading.Tasks;
using System.Collections.Generic;
using System.Linq;
using HearthstoneReplays.Events;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser.ReplayData.Meta;
using HearthstoneReplays.Parser.ReplayData.Meta.Options;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Enums;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using System.Collections;

#endregion

namespace HearthstoneReplays.Parser
{
    public enum StateType
    {
        GameState,
        PowerTaskList,
    }

    public class ParserState
    {
        public ParserState(StateType type, EventQueueHandler queueHandler, StateFacade stateFacade)
        {
            Logger.Log("Calling reset from ParserState constructor", type);
            this.StateType = type;
            this.StateFacade = stateFacade;
            this.NodeParser = new NodeParser(queueHandler, stateFacade, this.StateType);
        }

        public GameState GameState = new GameState();

        public NodeParser NodeParser;
        public StateFacade StateFacade;
        public HearthstoneReplay Replay { get; set; }
        public Game CurrentGame { get; set; }
        public GameData GameData { get; set; }
        public SendChoices SendChoices { get; set; }
        public Choices Choices { get; set; }
        public Options Options { get; set; }
        public Option CurrentOption { get; set; }
        public object LastOption { get; set; }
        public SubSpell CurrentSubSpell { get; set; }
        public int FirstPlayerEntityId { get; set; }
        public int CurrentPlayerId { get; set; }
        public ChosenEntities CurrentChosenEntites { get; set; }
        public bool Ended { get; set; }
        public int NumberOfCreates { get; set; }
        public bool ReconnectionOngoing { get; set; }
        public bool Spectating { get; set; }
        //public string FullLog { get; set; } = "";

        public StateType StateType { get; private set; }

        private Node _node;
        public Node Node
        {
            get { return _node; }
            set
            {
                if (value != _node)
                {
                    if (_node != null
                        // This works because Tag and TagChanges don't create new nodes
                        && (_node.Type == typeof(FullEntity)
                                || _node.Type == typeof(ShowEntity)
                                || _node.Type == typeof(ChangeEntity)
                    ))
                    {
                        EndAction();
                        if (_node.Type == typeof(ShowEntity))
                        {
                            GameState.ShowEntity(_node.Object as ShowEntity);
                        }
                        else if (_node.Type == typeof(FullEntity))
                        {
                            // In this case we just update the current state to whatever the game 
                            // tells us to
                            //if (ReconnectionOngoing)
                            //{
                            //    GameState.UpdateTagsForFullEntity(_node.Object as FullEntity);
                            //}
                            //else
                            //{
                            GameState.FullEntity(_node.Object as FullEntity, false);
                            //}
                        }
                        else if (_node.Type == typeof(ChangeEntity))
                        {
                            GameState.ChangeEntity(_node.Object as ChangeEntity);
                        }
                    }
                    if (_node != null && _node.Type == typeof(PlayerEntity))
                    {
                        NodeParser.CloseNode(_node, StateType);
                        GameState.PlayerEntity(_node.Object as PlayerEntity);
                    }
                    if (_node != null && _node.Type == typeof(GameEntity))
                    {
                        NodeParser.CloseNode(_node, StateType);
                        GameState.GameEntity(_node.Object as GameEntity);
                    }
                    if (_node != null && _node.Type == typeof(MetaData))
                    {
                        EndAction();
                    }
                    //HandleNodeUpdateEvent(_node, value);
                    this._node = value;
                }
            }
        }

        private Player _localPlayer;
        public Player LocalPlayer
        {
            get { return _localPlayer; }
        }

        public void SetLocalPlayer(Player value, DateTime timestamp, string data, bool sendEvent)
        {
            _localPlayer = value;
            value.IsMainPlayer = true;
            var players = getPlayers();
            var playerEntity = getPlayers().Find(player => player.PlayerId == value.PlayerId);
            playerEntity.IsMainPlayer = value.IsMainPlayer;
            if (sendEvent)
            {
                NodeParser.EnqueueGameEvent(new List<GameEventProvider> { GameEventProvider.Create(
                    timestamp,
                    "LOCAL_PLAYER",
                    () => new GameEvent
                    {
                        Type = "LOCAL_PLAYER",
                        Value = this._localPlayer
                    },
                    false,
                    new Node(null, null, 0, null, data)
            )});
            }
        }

        private Player _opponentPlayer;
        public Player OpponentPlayer
        {
            get { return _opponentPlayer; }
        }

        public void SetOpponentPlayer(Player value, DateTime timestamp, string data, bool sendEvent)
        {
            _opponentPlayer = value;
            value.IsMainPlayer = false;
            var playerEntity = getPlayers().Find(player => player.PlayerId == value.PlayerId);
            playerEntity.IsMainPlayer = value.IsMainPlayer;
            //var gameState = GameEvent.BuildGameState(this, StateFacade, GameState, null, null);
            if (sendEvent)
            {
                NodeParser.EnqueueGameEvent(new List<GameEventProvider> { GameEventProvider.Create(
                    timestamp,
                    "OPPONENT_PLAYER",
                    () => new GameEvent
                    {
                        Type = "OPPONENT_PLAYER",
                        Value = new {
                            OpponentPlayer = this._opponentPlayer,
                            //GameState = gameState,
                        }
                    },
                    false,
                    new Node(null, null, 0, null, data)
                )});
            }
        }

        public void Reset(StateFacade helper)
        {
            GameState.Reset(this);
            NodeParser.Reset(this, helper);
            Replay = new HearthstoneReplay();
            Replay.Games = new List<Game>();
            CurrentGame = new Game();
            this._localPlayer = null;
            this._opponentPlayer = null;
            // Prevent an EndAction from triggering
            _node = null;
            Node = null;
            GameData = null;
            SendChoices = null;
            Choices = null;
            Options = null;
            CurrentOption = null;
            LastOption = null;
            FirstPlayerEntityId = -1;
            CurrentPlayerId = -1;
            CurrentChosenEntites = null;
            Ended = false;
            ReconnectionOngoing = false;
            //FullLog = "";
            NumberOfCreates = 0;
            Logger.Log($"resetting game state, ended is now Ended={Ended}", this.StateType);

            _isBattlegrounds = null;
            _cachedPlayers = new List<PlayerEntity>();
        }

        public void PartialReset()
        {
            var gameEntity = CurrentGame.Data.Where(d => d is GameEntity).Select(d => d as GameEntity).FirstOrDefault();
            var playerEntities = getPlayers();
            var keysToKeep = new HashSet<int> { gameEntity.Id }; // Exemple : clés à conserver
            foreach (var player in playerEntities)
            {
                keysToKeep.Add(player.Id);
            }
            GameState.CurrentEntities = GameState.CurrentEntities
                .Where(kvp => keysToKeep.Contains(kvp.Key))
                .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
        }

        public void CreateNewNode(Node newNode)
        {
            NodeParser.NewNode(newNode, StateType);
        }

        public void EndAction()
        {
            var debug = Node.Type == typeof(FullEntity) && (Node.Object as FullEntity).Entity == 68;
            if (Node.Type != typeof(Game))
            {
                NodeParser.CloseNode(Node, StateType);
            }
        }

        public void EndCurrentGame()
        {
            Ended = true;
            Logger.Log($"Game Ended in type={StateType}", $"Ended={Ended}");
        }

        public void UpdateCurrentNode(params System.Type[] types)
        {
            while (Node?.Parent != null && types.All(x => x != Node.Type))
                Node = Node.Parent;
        }

        private List<PlayerEntity> _cachedPlayers = new List<PlayerEntity>();
        public List<PlayerEntity> getPlayers()
        {
            if (_cachedPlayers.Count > 0)
            {
                return _cachedPlayers;
            }
            List<PlayerEntity> players = new List<PlayerEntity>();
            var dataCopy = CurrentGame.Data.ToList();
            foreach (GameData x in dataCopy)
            {
                if (x is PlayerEntity) players.Add((PlayerEntity)x);
            }
            // Not yet, need further testing to make sure it doesn't break anything
            //if (players.Count > 0)
            //{
            //    _cachedPlayers = players;
            //}
            return players;
        }

        public PlayerEntity GetPlayerForController(int controllerId)
        {
            var players = getPlayers();
            foreach (var player in players)
            {
                if (player.PlayerId == controllerId) return player;
            }
            return null;
        }

        public void TryAssignLocalPlayer(DateTime timestamp, string data)
        {

            // In mercenaries, there are no HERO_ENTITY tags, because there are, well
            // no heroes.
            if (IsMercenaries())
            {
                if (getPlayers().Count == 3 && CurrentGame.ScenarioID == (int)ScenarioId.LETTUCE_PVP)
                {
                    CurrentGame.ScenarioID = (int)ScenarioId.LETTUCE_PVP_VS_AI;

                }
                
                var localPlayerPlayerId = -1;
                var opponentPlayerPlayerId = -1;

                // Doing things intelligently is very difficult, because of the "3-players" approach they have in the logs.
                // The way players are assigned seem to be consistent across games though, so it's in fact probably easier
                // to hard-code the mapping
                // Mercenaries PvE has 3 players. From what I've seen:
                // - The first player is the main player, but a "dummy" account? Maybe used to store mercs in some circumstances?
                // - The second player is the AI
                // - The third player is the "real" player account
                if (IsMercenariesPvE() && getPlayers().Count == 3)
                {
                    localPlayerPlayerId = getPlayers()[2].PlayerId;
                    opponentPlayerPlayerId = getPlayers()[1].PlayerId;
                }
                else
                {
                    // We assume that our heroes are always the first ones revealed
                    localPlayerPlayerId = CurrentGame
                        .FilterGameData(typeof(FullEntity))
                        .Where(d => d is FullEntity)
                        .Select(d => d as FullEntity)
                        .Where(d => d.GetTag(GameTag.LETTUCE_MERCENARY) == 1 && d.CardId?.Length > 0)
                        .FirstOrDefault()
                        ?.GetEffectiveController() ?? -1;
                    opponentPlayerPlayerId = CurrentGame
                        .FilterGameData(typeof(FullEntity))
                        .Where(d => d is FullEntity)
                        .Select(d => d as FullEntity)
                        // On PvE we know the CardId
                        .Where(d => d.GetTag(GameTag.LETTUCE_MERCENARY) == 1 && (d.CardId?.Length == 0 || d.GetZone() == (int)Zone.PLAY))
                        // When reconnecting, our own mercs can already be in play, so we need to make sure
                        // we're picking a different controller
                        .Where(d => d.GetEffectiveController() != localPlayerPlayerId)
                        .FirstOrDefault()
                        ?.GetEffectiveController() ?? -1;
                }

                foreach (PlayerEntity player in getPlayers())
                {
                    if (player.PlayerId == opponentPlayerPlayerId && data.Contains("PlayerID=" + opponentPlayerPlayerId))
                    {
                        var newPlayer = Player.from(player);
                        SetOpponentPlayer(newPlayer, timestamp, data, StateType == StateType.GameState);
                        return;
                    }
                    else if (player.PlayerId == localPlayerPlayerId && data.Contains("PlayerID=" + localPlayerPlayerId))
                    {
                        var newPlayer = Player.from(player);
                        FirstPlayerEntityId = player.Id;
                        SetLocalPlayer(newPlayer, timestamp, data, StateType == StateType.GameState);
                        return;
                    }

                }
                return;
            }

            // Only assign the local player once
            // For mercenaries this is a bit different, since we have 4 player ID assignments
            if (LocalPlayer != null && OpponentPlayer != null)
            {
                return;
            }

            // Names are not assigned right away, so wait until all the data is present to notify
            foreach (PlayerEntity player in getPlayers())
            {
                // SOme games against AI, eg Bob
                if (player.Name == null && player.AccountHi != "0")
                {
                    //Console.WriteLine("Player with no name: " + player);
                    return;
                }
                var playerEntityIdTag = player.Tags.Where(t => t.Name == (int)GameTag.HERO_ENTITY).First();
                if (playerEntityIdTag == null)
                {
                    return;
                }
            }

            //Console.WriteLine("Trying to assign local player");
            List<IEntityData> showEntities = CurrentGame.FilterGameData(typeof(ShowEntity))
                .Select(d => d as ShowEntity) // We need the HAND zone, because other effects can trigger before mulligan (like anomalies)
                .Where(d => d.GetZone() == (int)Zone.HAND)
                // Anomalies mess this up, as they trigger before mulligan, and are assigned first to a player
                // at random (apparenlty)
                //.Where(d => d.GetTag(GameTag.CREATOR) != GameState.GetGameEntity()?.Id)
                .Select(d => (IEntityData)d)
                .ToList();
            // Happens when facing Bob, or when reconnecting
            if (showEntities.Count == 0)
            {
                Logger.Log("No show entity, fallback to fullentity in hand", "");
                showEntities = CurrentGame
                    .FilterGameData(typeof(FullEntity))
                    .Select(d => d as FullEntity)
                    .Where(d => d.GetZone() == (int)Zone.HAND)
                    // Anomalies mess this up, as they trigger before mulligan, and are assigned first to a player
                    // at random (apparenlty)
                    .Where(d => d.GetTag(GameTag.CREATOR) != GameState.GetGameEntity()?.Id)
                    .Select(d => (IEntityData)d)
                    .ToList();
                if (showEntities.Count == 0)
                {
                    Logger.Log("No full entity in hand, fallback to fullentity", "");
                    showEntities = CurrentGame.FilterGameData(typeof(FullEntity)).Select(d => (IEntityData)d).ToList();
                }
            }
            foreach (IEntityData entity in showEntities)
            {
                if (entity.CardId != null && entity.CardId.Length > 0
                    && GetTag(entity.Tags, GameTag.CARDTYPE) != (int)CardType.ENCHANTMENT
                    // We do this because some cards are revealed when drawn (like Aranasi Bloodmother) and mess up with
                    // this logic
                    // We can't use zone=HAND here because of Battlegrounds
                    && GetTag(entity.Tags, GameTag.ZONE) != (int)Zone.DECK)
                {
                    int entityId = entity.Entity;
                    BaseEntity fullEntity = GetEntity(entityId);
                    int controllerId = fullEntity.GetEffectiveController();
                    //Console.WriteLine("Passed first step: " + entityId + ", " + fullEntity + ", " + controllerId);
                    foreach (PlayerEntity player in getPlayers())
                    {
                        if (player.GetEffectiveController() == controllerId)
                        {
                            var newPlayer = Player.from(player);
                            var playerEntityId = player.Tags.Where(t => t.Name == (int)GameTag.HERO_ENTITY).First().Value;
                            FullEntity playerEntity = CurrentGame.Data
                                .Where(d => d is FullEntity)
                                .Select(d => (FullEntity)d)
                                .Where(e => e.Id == playerEntityId)
                                .First();
                            newPlayer.CardID = playerEntity.CardId;
                            SetLocalPlayer(newPlayer, timestamp, data, StateType == StateType.GameState);
                        }
                    }
                    if (LocalPlayer != null)
                    {
                        foreach (PlayerEntity player in getPlayers())
                        {
                            if (player.Id == LocalPlayer.Id)
                            {
                                continue;
                            }
                            var newPlayer = Player.from(player);
                            var playerEntityId = player.Tags.Where(t => t.Name == (int)GameTag.HERO_ENTITY).First().Value;
                            FullEntity playerEntity = CurrentGame.Data
                                .Where(d => d is FullEntity)
                                .Select(d => (FullEntity)d)
                                .Where(e => e.Id == playerEntityId)
                                .First();
                            newPlayer.CardID = playerEntity.CardId;
                            SetOpponentPlayer(newPlayer, timestamp, data, StateType == StateType.GameState);
                        }
                        return;
                    }
                }
            }
            // Could not assign any player
            if (_localPlayer == null && _opponentPlayer == null)
            {
                Logger.Log("ERROR TO LOG: Could not assign local player " + data, string.Join(", ", getPlayers()?.Select(player => player.Name).ToList()));
            }
        }

        public int GetTag(List<Tag> tags, GameTag tag)
        {
            Tag ret = tags.FirstOrDefault(t => t.Name == (int)tag);
            return ret == null ? -1 : ret.Value;
        }

        public BaseEntity GetEntity(int id)
        {
            return CurrentGame.FilterGameData(typeof(FullEntity), typeof(PlayerEntity))
                .Select(data => (BaseEntity)data).ToList()
                .Where(e => e.Id == id)
                .FirstOrDefault();
        }

        private bool? _isBattlegrounds = null;
        public bool IsBattlegrounds()
        {
            if (_isBattlegrounds != null)
            {
                return this._isBattlegrounds.Value;
            }
            if (CurrentGame.GameType == -1)
            {
                return false;
            }
            this._isBattlegrounds = CurrentGame.GameType == (int)GameType.GT_BATTLEGROUNDS
                || CurrentGame.GameType == (int)GameType.GT_BATTLEGROUNDS_FRIENDLY
                || CurrentGame.GameType == (int)GameType.GT_BATTLEGROUNDS_DUO
                || CurrentGame.GameType == (int)GameType.GT_BATTLEGROUNDS_DUO_FRIENDLY;
            return this._isBattlegrounds.Value;
        }

        public bool IsBattlegroundsDuos()
        {
            return CurrentGame.GameType == (int)GameType.GT_BATTLEGROUNDS_DUO
                || CurrentGame.GameType == (int)GameType.GT_BATTLEGROUNDS_DUO_FRIENDLY;
        }

        public bool InRecruitPhase()
        {
            return GameState.GetGameEntity().GetTag(GameTag.BOARD_VISUAL_STATE) == 1;
        }

        public bool InCombatPhase()
        {
            return GameState.GetGameEntity().GetTag(GameTag.BOARD_VISUAL_STATE) == 2;
        }

        public bool IsMercenaries()
        {
            return IsMercenariesPvE() || new List<int>() {
                (int)GameType.GT_MERCENARIES_PVP,
            }.Contains(CurrentGame.GameType);
        }
        public bool IsMercenariesPvE()
        {
            return new List<int>() {
                (int)GameType.GT_MERCENARIES_AI_VS_AI,
                (int)GameType.GT_MERCENARIES_FRIENDLY,
                (int)GameType.GT_MERCENARIES_PVE,
                (int)GameType.GT_MERCENARIES_PVE_COOP,
            }.Contains(CurrentGame.GameType);
        }

        internal bool IsReconnecting(long gameSeedForCurrentLogs)
        {
            Logger.Log("Is reconnecting?", $"statType={StateType}, ended={Ended}, creates={NumberOfCreates}, spectating={Spectating}, gameSeed={CurrentGame?.GameSeed}, currentGameSeed={gameSeedForCurrentLogs}");
            return gameSeedForCurrentLogs != 0 
                && CurrentGame != null 
                && CurrentGame.GameSeed != 0 
                && CurrentGame.GameSeed == gameSeedForCurrentLogs;
            //return !Ended && NumberOfCreates >= 1 && !Spectating;
        }

        public void ClearActiveSubSpell()
        {
            var current = this.CurrentSubSpell;
            SubSpell parent = null;
            if (current?.Spell == null)
            {
                this.CurrentSubSpell = null;
            }
            while (current?.Spell != null)
            {
                parent = current;
                current = current.Spell;
            }
            if (parent != null)
            {
                parent.Spell = null;
            }
        }
    }
}